applications/MeshMovingApplication/custom_utilities/calculate_mesh_velocity_utility.cpp

```diff
@@ -0,0 +1,202 @@
+        << current_buffer_size << " < " << min_buffer_size << "!" << std::endl;
+
+    // set Generalized-Alpha specific Settings
+    if (mIntegrationMethod == generalized_alpha) {
+        KRATOS_ERROR_IF_NOT(rModelPart.HasNodalSolutionStepVariable(MESH_ACCELERATION))
+            << "The ModelPart does not have the variable \"MESH_ACCELERATION\" as "
+            << "nodal-solutionstepvariable!" << std::endl;
+
+        // default (Newmark)
+        double alpha_m = 0.0;
+        double alpha_f = 0.0;
+
+        // setting alpha_m and alpha_f
+        if (time_scheme == "generalized_alpha") {
+            alpha_m = Settings["alpha_m"].GetDouble();
+            alpha_f = Settings["alpha_f"].GetDouble();
+        } else if (time_scheme == "bossak") {
+            if (Settings.Has("alpha_m")) {
+                alpha_m = Settings["alpha_m"].GetDouble();
+            }
+            else {
+                alpha_m = -0.3; // default in Kratos
+            }
+        }
+
+        mbeta = std::pow((1.0 + alpha_f - alpha_m), 2) * 0.25;
+        mgamma = 0.5 + alpha_f - alpha_m;
+    }
+}
+
+CalculateMeshVelocityUtility::SizeType CalculateMeshVelocityUtility::GetMinimumBufferSize(
+    const std::string& rIntegrationMethod)
+{
+    return std::get<1>(GetMethod(rIntegrationMethod));
+}
+
+void CalculateMeshVelocityUtility::CalculateMeshVelocities()
+{
+    const double delta_time = mrModelPart.GetProcessInfo()[DELTA_TIME];
+    KRATOS_ERROR_IF(delta_time < 1.0e-24) << "Detected delta_time = 0! "
+        << "check if the time step is created correctly "
+        << "for the current time step" << std::endl;
+
+    if ( mIntegrationMethod == bdf1 ||
+         mIntegrationMethod == bdf2 ) {
+        CalculateMeshVelocitiesBDF(delta_time);
+    }
+    else {
+        CalculateMeshVelocitiesGeneralizedAlpha(delta_time);
+    }
+
+    mrModelPart.GetCommunicator().SynchronizeVariable(MESH_VELOCITY);
+}
+
+void CalculateMeshVelocityUtility::CalculateMeshVelocitiesBDF(const double DeltaTime)
+{
+    const int num_local_nodes = mrModelPart.GetCommunicator().LocalMesh().NumberOfNodes();
+    const auto nodes_begin = mrModelPart.GetCommunicator().LocalMesh().NodesBegin();
+
+    const double coeff = 1 / DeltaTime;
+
+    switch(mIntegrationMethod) {
+        case bdf1 : {
+            #pragma omp parallel for
+            for (int i=0; i<num_local_nodes; i++) {
+                const auto it_node  = nodes_begin + i;
+                auto& r_mesh_v0       = it_node->FastGetSolutionStepValue(MESH_VELOCITY);
+                const auto& r_mesh_u0 = it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT);
+                const auto& r_mesh_u1 = it_node->FastGetSolutionStepValue(MESH_DISPLACEMENT, 1);
+                noalias(r_mesh_v0) = r_mesh_u0 - r_mesh_u1;
+                r_mesh_v0 *= coeff;
+            }
+            break;
+        }
+        case bdf2 : {
+            const double c1 = 1.50 * coeff;
```

Comment:
guys this is a blocker! it does not work in the case of variable time steps!!

Reply:
@RiccardoRossi this has been like this for years :)
See the current master

As I wrote, this will be refactored once #3475 is available, then this will work automatically

Reply:
solved as requested